### ISSUE TITLE
[ASM] IncrementalBuilder with JavaLanguage does not work with Inner Classes

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-extras#30]
Having a Java Class like

```
package mypack;
public class FakeIssue60GrammarAccess {
	
	private final GreetingElements pGreeting;
	
	public FakeIssue60GrammarAccess() {
		this.pGreeting = new GreetingElements();
	}
	
	public class GreetingElements {

		public String getHelloKeyword_0() { return null; }
		
	}
	
	public GreetingElements getGreetingAccess() {
		return pGreeting;
	}

}	
```

and a Xtend class like

```
package mypack

import mypack.FakeIssue60GrammarAccess$GreetingElements

class Demo {
	
	extension mypack.FakeIssue60GrammarAccess
	
	GreetingElements elements

	def void format() {
		greetingAccess.helloKeyword_0
		println(elements)
	}
	
}
```

then the compilation with the incremental builder e.g. xtend-gradle-plugin fails with `ERROR:The method getGreetingAccess() from the type FakeIssue60GrammarAccess refers to the missing type Object `

see https://github.com/xtext/xtext-gradle-plugin/issues/60

**Comment by cdietrich** [comment]
This is the Analysis of the Problem

- asm creates classes with uris like `java:/Objects/org.eclipse.xtend.core.tests.build.Outer$Inner$InnerMost#org.eclipse.xtend.core.tests.build.Outer$Inner$InnerMost`
- index entries look like `org.eclipse.xtend.core.tests.build.Outer..Inner.InnerMost`
- `IndexedJvmTypeAccess` uses the $URI to do the index lookup and will not find anything.

@szarnekow and i assume that we have to iterate over all possibilities there. (replacing `$`with `.`in all combinations.

@svenefftinge @JanKoehnlein what do you think.

**Comment by cdietrich** [comment]
Provided a prototype impl and some test (here and at https://github.com/eclipse/xtext-xtend/tree/cd_extras_issue30

Feedback and input is welcome. At InnerClassNameVariants i currently have 2 impls that (should) do the same. i dont know which one i (dis)like more.

**Comment by cdietrich** [comment]
feedback in welcome. all the rebasing drives me nuts

**Comment by svenefftinge** [comment]
Thx christian, for all the analysis. I think there is something wrong with how the java language contributes to the index, because this issue only seems to occur in that scenario.
I hope I can find time to look at this ASAP.

**Comment by cdietrich** [comment]
i debugged that as well, the visit methods build the names this way

**Comment by cdietrich** [comment]
i had a discussion with @szarnekow  on that. he meant

> wenn asm die klassen anschaut, ist nicht bekannt, ob ein nested class referneziert ist, oder ein top-level-class

thus i went the way to address this in the access

**Comment by svenefftinge** [comment]
Ok, in that case being more flexible with incoming names seems to be the best solution.
I like the stream-based approach a bit more.

**Comment by svenefftinge** [comment]
A PR would be awesome! :)

**Comment by cdietrich** [comment]
Fixed in Xtext 2.11
